### ISSUE TITLE
Another idea for the menu and the display of things

### ISSUE THREAD

**Issue by rwobben** [rwobben/paytime#36]
Hello, 

Im not complete happy with the menu. it's getting bigger and bigger. 
Also the tabular layout is not appealing and have not the proffessional look that I was hoping for. 

I think both can come back if I know more about Haskell, expecially functors and that sort of things.

Can we come up with a better alternative at the moment 
I have looked how hledger is doing things but that is also above my head. 
Hledger uses the states. So again monads :(

**Comment by GetContented** [comment]
Ok, so I'm not sure what this issue is about?

**Comment by rwobben** [comment]
LIke I said . The menu we have in main is becoming bigger and bigger,
LIke you said earlier. Then the code smells.  But I cannot figure out a way to make the menu smaller with the knowlegde I have.  So Im looking for a alternative. 

The same way with displaying when a user entered a new account. or a transaction.
The more I see it, the more I do not like it.  In my oponion it's only two rows of numbers. 
I like to think also about a way to make this look better. 

Also I wonder if we must continue with a cli-version. No modern accounting system works with a cli anymore. They are all web - based.

**Comment by GetContented** [comment]
Well, the size of functions is a general code smell... 

Trust me that you don't want to start doing web stuff!

You need to have done a larger chunk of Haskell before we can tackle web stuff. This whole project is just an exercise for you to learn, not a finished project. It's a toy for you to sharpen your teeth on.

**Comment by GetContented** [comment]
When you say "the menu we have" - which function are you talking about? For example, the main function has a lot of opportunity for refactoring. Is that what you mean?

**Comment by rwobben** [comment]
I know web stuff is way of my lequae. 

That is why im looking for a way to make things work. But a way I can live with it.
Yep, the main menu is mainly the menu.

**Comment by GetContented** [comment]
yes, but which function are you talking about? Are you talking about the function called `main` here?

**Comment by GetContented** [comment]
Personally I really dislike the folder structure that is currenly being used. We seem to have "cli" stuff in the same folders as the pure functions, but they should probably be in separate folders.

**Comment by rwobben** [comment]
With the menu we are talking about the main function, yes.

**Comment by rwobben** [comment]
??? you mean you dislike to be everything in one file

**Comment by GetContented** [comment]
Ok yes, `main` can be refactored pretty easily.

I don't much like how there's an Accounts folder which has cli.hs and lib.hs in it... this doesn't make much sense to me. I think there should be one place for accounts library code, and you would layer the cli stuff on top of that, if you even need to pull that into different files.

**Comment by GetContented** [comment]
The first step in refactoring main is to notice the repeating parts. Which parts of it are repeated? can you notice these patterns?

**Comment by rwobben** [comment]
IN the new one everything is in one file. I only have a Lib.hs which contains everything. 
The layout which you talk about was not working.  So I refractored all the common parts now to different functions and put it all in lib.hs

**Comment by rwobben** [comment]
in main there are two pieces that come back often 

This piece : 

     xxxx <- getline 
     hFlush stdout
                      choice<- getLine
                      case choice  of 

and this piece : 

   newAccounts <-  xxxx  accounts
                              runLoop newAccounts

**Comment by GetContented** [comment]
cool! yeah...

**Comment by GetContented** [comment]
actually we can probably pull each of the `menuChoice` case consequences out into their own `IO ()` values, don't you think?

**Comment by rwobben** [comment]
That is a way I have thought about but could not make it work

**Comment by rwobben** [comment]
or doing it this way : https://github.com/pjones/byline/blob/master/examples/menu.hs 
Then we ccan make a difference with colors between the menu and the output of a function.

**Comment by GetContented** [comment]
Well we could easily take one piece at a time.

Let's see... how about this:

            "AC" ->do 
                      let canDelete = length accounts > 0
                      putStrLn $ ""
                      putStrLn $ "subMenu Accounts \n"
                      putStrLn $"CA: Create Accounts.\n" ++ (if canDelete then "\nDA: Delete Account." else "")
                      putStr $ "Your Choice: "
                      hFlush stdout
                      choice<- getLine
                      case choice  of 
                          "CA" -> do
                              newAccounts <- doCreateAccounts accounts
                              runLoop newAccounts
                          "DA" -> do 
                               newAccounts <- doDeleteAccounts accounts
                               runLoop newAccounts


We know the return type of that must be IO ().

So we can extract that into a definition

**Comment by GetContented** [comment]
The trouble with using that example is that we have some state we need to pass around between loops of the main function....

**Comment by rwobben** [comment]
?? I do not see any state in that code

**Comment by GetContented** [comment]
the accounts is the state.

**Comment by rwobben** [comment]
of course, this function quits after a choice and there are all other choices.

oke, lets continue with the menu

**Comment by GetContented** [comment]
If you *are* going to extract the accounts menu portion into a separate function, you will need to pass all required context in... such as the runLoop function itself, and the accounts.

But this becomes a little ugly, really.

This is one reason not to tackle the UI portion of the app before you know a certain amount of Haskell - because complex UI involves monads.

It really *should* be done with the state monad to clean it up, and as you say, we're not going to be using the CLI in our final version. :) Maybe we should abandon this project for now.

**Comment by rwobben** [comment]
I think the same. 
Or I could delete it and begin with a fresh one  when I finisched the book.
Then we could think about a structure that works. 
I do not know if the state monad is discussed in the book. 

I want to thank you for all your lessons and patience.
I will continue with the Haskell book. the challenges costs enought time to solve.

**Comment by GetContented** [comment]
Yeah, the state monad is discussed in the book.

Yeah, it's worthwhile just sticking through with the book.

If you want to thank me, buy a copy of my book, and get as many people as you can to buy it :) Assuming you think it's good, that is. http://happylearnhaskelltutorial.com

**Comment by GetContented** [comment]
And yeah, once you've finished it maybe you can tackle using Yesod or Scotty or Snap to build yourself a web-based accounting package. You won't need me to help you then :)
